base/util.go

```diff
@@ -590,27 +591,117 @@ func IsFilePathWritable(fp string) (bool, error) {
 	return true, nil
 }
 
-// Replaces sensitive data from the URL query string with ******.
+// SanitizeRequestURL will return a sanitised string of the URL by:
+// - Tagging mux path variables.
+// - Tagging query parameters.
+// - Replacing sensitive data from the URL query string with ******.
 // Have to use string replacement instead of writing directly to the Values URL object, as only the URL's raw query is mutable.
-func SanitizeRequestURL(requestURL *url.URL) string {
-	urlString := requestURL.String()
+func SanitizeRequestURL(req *http.Request) string {
+	urlString := SanitizeRequestURLQueryParams(req.URL)
+
+	if RedactSystemData || RedactMetadata || RedactUserData {
+		tagQueryParams(req, &urlString)
+		tagPathVars(req, &urlString)
+	}
+
+	return urlString
+}
+
+// redactedPathVars is a lookup map of path variables to redaction types.
+var redactedPathVars = map[string]string{
+	"docid":   "UD",
+	"attach":  "UD",
+	"name":    "UD",
+	"channel": "UD",
+
+	"db":        "MD",
+	"newdb":     "MD",
+	"ddoc":      "MD",
+	"view":      "MD",
+	"sessionid": "MD",
+}
+
+// tagPathVars will tag all redactble path variables in the urlString for the given request.
+func tagPathVars(req *http.Request, urlString *string) {
+	if urlString == nil {
+		return
+	}
+
+	str := *urlString
+	pathVars := mux.Vars(req)
+
+	for k, v := range pathVars {
+		switch redactedPathVars[k] {
+		case "UD":
+			str = strings.Replace(str, v, UD(v).Redact(), 1)
+		case "MD":
+			str = strings.Replace(str, v, MD(v).Redact(), 1)
+		case "SD":
+			str = strings.Replace(str, v, SD(v).Redact(), 1)
+		}
+	}
+
+	*urlString = str
+}
+
+// redactedQueryParams is a lookup map of query params to redaction types.
+var redactedQueryParams = map[string]string{
+	"channels": "UD",
+	"doc_ids":  "UD",
+	"docid":    "UD",
+
+	"since":     "MD",
+	"rev":       "MD",
+	"open_revs": "MD",
```

Comment:
I know it's not a pressing issue for now, but I'm on the fence about some of these, in particular `since`. Happy to push out the conversation as `MD` isn't actually required yet, just flagging it up.

Reply:
I agree with you - any replication internals (since, revs, etc) shouldn't be subject to redaction.

Reply:
As far as I can tell, and unless mistaken, these values sit squarely in the metadata category defined in [Vulcan's Log Redaction spectrum](https://docs.google.com/document/d/1rbQsiDpKspJpzZspAMxjvU19cHEtvEFeGQ_jLmFJ4bM/edit#heading=h.q6wzxxq9l7kg)? Perhaps it does require further discussion.

> [2] Redaction Spectrum 
> Log files means files collected by the cb_collect info process in Couchbase for diagnostics and troubleshooting purposes.
> 
> User data is data that is stored into Couchbase by the application user account 
> * Key and value pairs in JSON documents, or the key exclusively
> * Application/Admin usernames that identify the human person 
> * Query statements included in the log file collected by support that leak the document fields (Select floor_price from stock).
> * Names and email addresses asked during product registration and alerting
> * Usernames
> * Document xattrs
> 
> Metadata is logical data needed by Couchbase to store and process User data.  
> * Cluster name
> * Bucket names
> * DDoc/view names
> * View code
> * Index names
> * Mapreduce Design Doc Name and Definition (IP)
> * XDCR Replication Stream Names
> * And other couchbase resource specific meta data 
> 
> System data is data from other parts of the system Couchbase interacts with over the network
> * IP addresses
> * IP tables
> * Hosts names
> * Ports
> * DNS topology